cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,427 @@
+      switch_to_dynamic(n);
+    }
+  }
+
+  void clear() {
+    this->destroy_storage(data_, size_);
+    size_ = 0;
+  }
+
+ private:
+  void switch_to_dynamic(size_t new_capacity) {
+    dynamic_capacity_ = new_capacity;
+    data_ = new storage_type[new_capacity];
+    this->move_storage(static_data_, data_, size_);
+  }
+
+  void reallocate_dynamic(size_t new_capacity) {
+    assert(new_capacity >= size_);
+    auto new_data = new storage_type[new_capacity];
+    this->move_storage(data_, new_data, size_);
+    delete[] data_;
+    dynamic_capacity_ = new_capacity;
+    data_ = new_data;
+  }
+};
+
+template <typename T, size_t N, typename Base>
+class StaticVectorImpl : public Base {
+ private:
+  using Base::size_;
+
+  T* item_ptr(size_t i) { return this->ptr_at(this->data_ptr(), i); }
+
+  constexpr const T* const_item_ptr(size_t i) const {
+    return this->ptr_at(this->const_data_ptr(), i);
+  }
+
+ public:
+  using size_type = size_t;
+  using difference_type = ptrdiff_t;
+  using value_type = T;
+  using pointer = T*;
+  using const_pointer = const T*;
+  using reference = T&;
+  using const_reference = const T&;
+  using iterator = T*;
+  using const_iterator = const T*;
+
+  constexpr StaticVectorImpl() noexcept = default;
+
+  StaticVectorImpl(StaticVectorImpl&& other) noexcept {
+    this->move_from(std::move(other));
+  }
+
+  StaticVectorImpl& operator=(StaticVectorImpl&& other) noexcept {
+    if (&other != this) {
+      this->destroy();
+      this->move_from(std::move(other));
+    }
+    return *this;
+  }
+
+  StaticVectorImpl(const StaticVectorImpl& other) {
+    const size_t n = other.size_;
+    this->bump_size(n);
+    auto* src = other.const_item_ptr(0);
+    auto* dest = this->item_ptr(0);
+    for (size_t i = 0; i < n; ++i) {
+      new (&dest[i]) T(src[i]);
+    }
+  }
+
+  StaticVectorImpl& operator=(const StaticVectorImpl& other) noexcept {
+    if (&other == this) {
+      return *this;
+    }
+    const size_t n = other.size_;
+    const size_t old_size = size_;
+    auto* src = other.const_item_ptr(0);
+    if (n > old_size) {
+      this->bump_size(n);
+      auto* dest = this->item_ptr(0);
+      for (size_t i = 0; i < old_size; ++i) {
+        dest[i] = src[i];
+      }
+      for (size_t i = old_size; i < n; ++i) {
+        new (&dest[i]) T(src[i]);
+      }
+    } else {
+      auto* dest = this->item_ptr(0);
+      for (size_t i = 0; i < n; ++i) {
+        dest[i] = src[i];
+      }
+      for (size_t i = n; i < old_size; ++i) {
+        dest[i].~T();
+      }
+      this->reduce_size(n);
+    }
+    return *this;
+  }
+
+  explicit StaticVectorImpl(size_t count) {
+    this->bump_size(count);
+    auto* p = this->item_ptr(0);
+    for (size_t i = 0; i < count; ++i) {
+      new (&p[i]) T();
+    }
+  }
+
+  StaticVectorImpl(size_t count, const T& value) {
+    this->bump_size(count);
+    auto* p = this->item_ptr(0);
+    for (size_t i = 0; i < count; ++i) {
+      new (&p[i]) T(value);
+    }
+  }
+
```

Comment:
I think for usability we'd also want (possibly implicit) construction and assignment from `std::vector`. I think this would make replacement of `ArrayData::buffers` less painful